File: apps/web/src/lib/billing.ts
import { api } from "./api";

export type Plan = "free" | "reader" | "scholar" | "family";
export type Cycle = "monthly" | "yearly";
export type SubscriptionStatus =
  | "inactive"
  | "trialing"
  | "active"
  | "past_due"
  | "canceled"
  | "unpaid";

export interface Quota {
  pages_per_month: number;
  max_pages_per_book: number;
  quizzes_per_book: number;
  profiles: number;
  chat_with_citations: boolean;
  annotated_export: boolean;
  priority_queue: boolean;
  family_safe_mode: boolean;
  literary_translation: boolean;
}

export interface Usage {
  period_start: string | null;
  pages_uploaded: number;
  quizzes_generated: number;
}

export interface Subscription {
  plan: Plan;
  cycle: Cycle | null;
  status: SubscriptionStatus;
  current_period_start: string | null;
  current_period_end: string | null;
  trial_end: string | null;
  cancel_at_period_end: boolean;
  canceled_at: string | null;
  has_stripe_customer: boolean;
  quota: Quota;
  usage: Usage;
}

export async function getSubscription(): Promise<Subscription> {
  return api<Subscription>("/billing/me");
}

export async function startCheckout(opts: {
  plan: Exclude<Plan, "free">;
  cycle: Cycle;
  applyFirstMonthDiscount?: boolean;
}): Promise<{ url: string }> {
  return api<{ url: string }>("/billing/checkout", {
    method: "POST",
    body: {
      plan: opts.plan,
      cycle: opts.cycle,
      apply_first_month_discount: opts.applyFirstMonthDiscount ?? false,
    },
  });
}

export async function startPortal(): Promise<{ url: string }> {
  return api<{ url: string }>("/billing/portal", { method: "POST" });
}

const PLAN_NAME: Record<Plan, string> = {
  free: "Free",
  reader: "Reader",
  scholar: "Scholar",
  family: "Family",
};

export function planName(p: Plan): string {
  return PLAN_NAME[p];
}

export function isUnlimited(n: number): boolean {
  return n >= 1_000_000;
}

File: apps/api/app/schemas/billing.py
"""Billing schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class CheckoutRequest(BaseModel):
    plan: Literal["reader", "scholar", "family"]
    cycle: Literal["monthly", "yearly"]
    apply_first_month_discount: bool = False


class CheckoutResponse(BaseModel):
    url: str


class PortalResponse(BaseModel):
    url: str


class QuotaRead(BaseModel):
    pages_per_month: int
    max_pages_per_book: int
    quizzes_per_book: int
    profiles: int
    chat_with_citations: bool
    annotated_export: bool
    priority_queue: bool
    family_safe_mode: bool
    literary_translation: bool
    # Page after which a Free reader is paywalled on seed books. The frontend
    # reads this from /billing/me to fire the upgrade modal — paid plans get
    # the UNLIMITED sentinel and never trip the gate. 1-indexed.
    seed_book_page_cap: int


class UsageRead(BaseModel):
    period_start: datetime | None
    pages_uploaded: int
    quizzes_generated: int


class SubscriptionRead(BaseModel):
    plan: Literal["free", "reader", "scholar", "family"]
    cycle: Literal["monthly", "yearly"] | None
    status: Literal[
        "inactive", "trialing", "active", "past_due", "canceled", "unpaid"
    ]
    current_period_start: datetime | None
    current_period_end: datetime | None
    trial_end: datetime | None
    cancel_at_period_end: bool
    canceled_at: datetime | None
    has_stripe_customer: bool
    quota: QuotaRead
    usage: UsageRead
